AWS: Caching layer for volumes & instances

justinsb opened the pull request:
AWS quota is request-based, not based on complexity of the calls.

Therefore we move volume & instance polling to a unified cache, where
any cache miss triggers a full refresh.  This should be much lighter on
requests that previous per-volume / per-instance polling.

This cache can be used more widely, but the challenge is keeping the
change from becoming overwhelming.  Initially therefore, we only use it
for the volume code.

```release-note
AWS: Introduce volume & instance caching into volumes code
```

justinsb committed:
AWS: Caching layer for volumes & instances

AWS quota is request-based, not based on complexity of the calls.

Therefore we move volume & instance polling to a unified cache, where
any cache miss triggers a full refresh.  This should be much lighter on
requests that previous per-volume / per-instance polling.

This cache can be used more widely, but the challenge is keeping the
change from becoming overwhelming.  Initially therefore, we only use it
for the volume code.

k8s-reviewable commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.kubernetes.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/41698)
<!-- Reviewable:end -->

justinsb commented:
cc @gnufied @jingxu97 

Pretty similar to some of my prior efforts (much of the same code), but really focused on instances & volumes.

justinsb commented:
BTW I am thinking we might want both bulk polling & caching ... bulk polling because it would be nice to know where a volume is attached I think.  I _think_ we can make use of that information.

justinsb commented:
@k8s-bot cvm gke e2e test this

gnufied commented on pkg/cloudprovider/providers/aws/instancecache.go:
```diff
@@ -0,0 +1,308 @@
+	c.mutex.Lock()
+	defer c.mutex.Unlock()
+
+	if c.latest != nil && c.latest != previous {
```
Why do we need this? won't the lock prevent concurrent `refreshAll`?

justinsb replied:
The way I think about it is that two refreshAll requests come in at the same time, with the same `previous`.  One holds at the mutex lock while the other refreshes.  Then the second thread acquires the lock, but we have already updated.  So we want to skip the second refresh.

gnufied commented on pkg/cloudprovider/providers/aws/instancecache.go:
```diff
@@ -0,0 +1,308 @@
+		} else {
+			if byNodeName[nodeName] != nil {
+				glog.Errorf("detected two instances with the same NodeName: %q", nodeName)
+			} else {
```
Shouldn't we simply replace the instance even if another instance with same name was found?

justinsb replied:
We created byNodeName as an empty map a few lines above.  So this is when we saw two instances with the same nodename.  Choosing between them is arbitrary, but we need to alert that something has gone wrong.

We _could_ actually return an error here or even panic - things are not going to go well from this point on - we're going to be choosing instances randomly.

gnufied commented on pkg/cloudprovider/providers/aws/aws.go:
```diff
@@ -1705,25 +1687,11 @@ func (c *Cloud) GetVolumeLabels(volumeName KubernetesVolumeID) (map[string]strin
 func (c *Cloud) DiskIsAttached(diskName KubernetesVolumeID, nodeName types.NodeName) (bool, error) {
-	_, instance, err := c.getFullInstance(nodeName)
+	// NeverCache - but TODO remove this function
+	_, instance, err := c.getFullInstance(NeverCache, nodeName)
```
I think one problem here is, this will pretty much always trigger a full instance cache refresh  - as it is currently implemented.

I think we should leave room for ability to read-through cache in certain cases and may be not force a full refresh because I am wondering if we would end up making more API requests than before.

justinsb replied:
This will, but hence the TODO to remove the function entirely.  Though I think your argument does apply to `DetachDisk`

The surprising thing is that a full refresh should be as quota-expensive as a single volume request.  That's not true with pagination, but DescribeVolumes actually returns all instances.

There are other concerns as well than just quota - we pull down a lot more data over the wire, for example.  But quota and correctness are much more important right now.

I'm not opposed to a single volume fetch, I just don't think we know we need it yet.

I could see us starting to do an LRU type fetching strategy, on the grounds that every time we fetch 1 volume we should actually fetch 50, but I suspect also that is simply a lot of complexity for not a lot of gain.

justinsb committed:
Simplify caching layer

Reduce usage of caching to bare minimum, document the caching behaviour.

justinsb commented:
@jingxu97 @gnufied what do you think - I've added documentation, and the _usage_ of the cache layer is vastly simplified.  I left the structure of the caches (and added comments), but I removed a lot of the functions that are no longer needed.

I would really like to get this into 1.6 - I think it's our only shot really at getting to the root of these request quota problems.  It missed 1.5 and it bit us hard, and we just have to get it in now.  I think the model is simple enough that we can understand it, and I can gradually move more functions to use the cache.

justinsb commented:
@k8s-bot cvm gce e2e test this
@k8s-bot gce etcd3 e2e test this

jingxu97 commented on pkg/cloudprovider/providers/aws/instancecache.go:
```diff
@@ -0,0 +1,328 @@
+	InstanceType     string
+	SubnetID         string
+
+	cache *instanceCache
```
Is this instanceCache necessary?

jingxu97 commented on pkg/cloudprovider/providers/aws/instancecache.go:
```diff
@@ -0,0 +1,328 @@
+	if snapshot != nil {
+		instance := snapshot.findCachedInstance(i.ID)
+		if instance != nil {
+			state := i.findStateIfCached(cachePolicy)
```
why not directly use i.state after checking the cachePolicy?

jingxu97 commented on pkg/cloudprovider/providers/aws/instancecache.go:
```diff
@@ -0,0 +1,328 @@
+		return nil, cloudprovider.InstanceNotFound
+	}
+
+	state := i.findStateIfCached(nil)
```
can use i.state directly?

jingxu97 commented on pkg/cloudprovider/providers/aws/instancecache.go:
```diff
@@ -0,0 +1,328 @@
+
+// findCachedInstance does a lookup of the instance by id, returning nil if not found.
+// It does not trigger a cache refresh.
+func (s *instanceCacheSnapshot) findCachedInstance(instanceID awsInstanceID) *cachedInstance {
```
is this necessary?

justinsb replied:
Yes - used in DescribeInstance around like 98 & 114

jingxu97 commented:
Current this PR implements two types of cache, volumeCache and instanceCache. In each cache struct, there is a latest snapshot which contains a list of items cached and the timestamp showing when the cache was last refreshed. For each cached item, it contains details (state) about the item and also an timestamp indicating when this item is last updated. 

A cache policy can be configured with different validity period. It is used to check whether the cache is considered valid or not. There are two places to use cache policy. First one is when finding the item from the snapshot and the policy checks the snapshot timestamp. The second one is when retrieving the state of of the cached item (DescribeItem) and the policy checks the item state timestamp. Current design is if the cachePolicy checking fails, it triggers a refresh which will request all instances or volumes information and update the cache at once.  Two predefined policy: NeverCached and NeverRefresh
  
In asw, two functions are changed to use this cached information.

1. During volume attaching/detaching, waitForAttachmentStatus keeps checking whether the specified volume status is changed to the expected state. 
 
- **Original function**: for each volume, it pulls the status from cloud provider with exponential backoff. If it takes average M pulls to reach the expected volume status, the cost is N*M requests for N volumes trying to attach. Each request will return the response with single volume information S bytes. So the total data transferred would be N*M*S bytes
-**New function:** for each volume, it first tries to get the cached volume from the latest snapshot with NeverRefresh (it will trigger a refreshAll when the volume is not in the cache at all). Then it tries to get the state (DescribeItem) of the item with cache policy of 10 second valid period. Given the same example as above, there are N volumes attachment around the same time. If all of them are new volumes, in the worst case, there will be N refreshAll triggered when getting the cached volume from the snapshot. During the polling period, one refreshAll will be probably triggered for all N volumes since each refreshAll will get all the volumes information if these N volumes are querying the status within 10 second. If N is big, there might be multiple refreshAll triggered. If there are total W volumes, each request will cost S*W bytes data transferred. So roughly, during this process, there will be total N + M requests issued, the total data transferred would be (N+M)*S*W during N volumes being attached. (normally W >> N & M)
-**Optimization suggestion:** During the volume first attachment, there will be multiple refreshAll triggered about the same time and most of the information transferred are the exactly the same. So when snapshot is not available, it makes sense to trigger refreshAll. After that, if an item is not in the cache, we can only request the specific volume that is not in the cache. The total number of request is still N for N newly added volumes, but data transferred would be much reduced to N*S. 

2. Volume manager periodically checks whether volumes are still attached to a node. 
-**Original function**: for each node, it sends a request to get node instances from cloud provider. If there are total N nodes, there will be N requests. Current polling period is set to 1 minute, so the cost is N requests/min. The data transferred would be N*S/min 
-**New function**: for each node, it gets the cached instance information from the last snapshot with 15 second valid cache period. Since polling period is 1 minute, so cached information will be expired and trigger a refreshAll. For N node, each time there will be only 1 request, so the cost would be 1 request/min.  If the cluster has W nodes, data transferred would be W*s/min. (normally W >> N)
-**Optimization suggestion**: the list of volumes volume manager tries to check periodically is mostly quite stable. Instead of refreshing all, we can send a request only contains the nodes in the list so the data transferred could be reduced largely. But this would involve some API changes such as VolumesAreAttached in volume manager.

gnufied commented on pkg/cloudprovider/providers/aws/volumecache.go:
```diff
@@ -0,0 +1,265 @@
+			ci.state = volumeDetails
+			ci.stateTimestamp = now
+			ci.mutex.Unlock()
+		}
```
Why these locks are here btw? Is this to prevent from some other piece of code accidently setting `state` of `CachedInstance`? 

But if another goroutine already has access to a pointer to `ci` then this lock in itself won't prevent the mutation of cached instance.

justinsb replied:
findStateIfCached reads the state & stateTimestamp values.  It is also protected by the same mutex.

gnufied commented on pkg/cloudprovider/providers/aws/instancecache.go:
```diff
@@ -0,0 +1,328 @@
+			ci.mutex.Lock()
+			ci.state = i
+			ci.stateTimestamp = now
+			ci.mutex.Unlock()
```
Same as below.

jingxu97 commented on pkg/cloudprovider/providers/aws/volumecache.go:
```diff
@@ -0,0 +1,265 @@
+// It accepts the previous snapshot, and will not refresh if a concurrent request has triggered a concurrent refresh.
+func (c *volumeCache) refreshAll(previous *volumeCacheSnapshot) (*volumeCacheSnapshot, error) {
+	c.mutex.Lock()
+	defer c.mutex.Unlock()
```
refreshAll might takes a little while to finish. During this time, cache will be locked totally, so other operations will be blocked? I think during ec2 describeInsances or describeVolumes, reading cache  should not be block?

jingxu97 commented on pkg/cloudprovider/providers/aws/volumecache.go:
```diff
@@ -0,0 +1,265 @@
+	snapshot := i.cache.latestSnapshot()
+
+	if snapshot != nil {
+		instance := snapshot.findCachedVolume(i.ID)
```
what is the purpose of this findCachedVolume? a cachedVolume must exist, otherwise this function is not going to be called?

zmerlynn commented on pkg/cloudprovider/providers/aws/cache.go:
```diff
@@ -0,0 +1,43 @@
+/*
+Copyright 2014 The Kubernetes Authors.
```
+1

zmerlynn commented on pkg/cloudprovider/providers/aws/volumecache.go:
```diff
@@ -0,0 +1,265 @@
+/*
+Copyright 2014 The Kubernetes Authors.
```
Ditto on year.

justinsb replied:
Technically I've probably been trying since 2014, but point taken ;-)

zmerlynn reviewed (commented):
Generally LGTM, a couple of nits. I'm a little leery to drop in a cache layer this late in the release - it seems easy 1.7 fodder to make sure we have burn-in time, and I'm not sure we're bleeding enough to justify rushing the testing.

ethernetdan commented:
Removing PR from 1.6 milestone since code freeze has begun:
* If this is a bug fix, please label as `kind/bug` and reapply the milestone.
* If this is a feature for 1.6, an [exception](https://github.com/kubernetes/features/blob/master/EXCEPTIONS.md) should be filed for it before 11:00PM PST on 3/2. Note that most exceptions to code freeze are not accepted.
* If this is a feature for 1.7, feel free to add the 1.7 milestone to the PR.

/cc @kubernetes/release-team @kubernetes/kubernetes-release-managers

calebamiles commented:
Moving this issue to the 1.7 milestone. Sorry @justinsb. Please file an [exception](https://github.com/kubernetes/features/blob/master/EXCEPTIONS.md) if you would like to lobby for merge in the 1.6 release. Thanks.

**cc: @ethernetdan, @kubernetes/kubernetes-release-managers**

hjacobs commented:
@justinsb I just discussed this with my Zalando colleagues, we will try the changes out (applied to 1.6.2) and provide feedback. Thanks for your work!

mikkeloscar commented:
@justinsb as @hjacobs said we can test it out. But would it be possible to rebase on master?

justinsb committed:
Update copyright

justinsb committed:
Merge fixes

justinsb commented:
I'm going to go ahead and close this - we can revisit post cloudprovider-split-out.
